syntax = "proto2";
import "net.proto";

package ncode.dfs;

message PBConstraint {
  enum Type {
    UNKNOWN = 0;
    DUMMY = 1;
    AVOID_EDGE = 2;
    AVOID_VERTEX = 3;
    AND = 4;
    OR = 5;
    AVOID_PATH = 6;
    DEPREF_EDGE = 7;
    DEPREF_VERTEX = 8;
    VISIT_EDGE = 9;
    VISIT_VERTEX = 10;
    VISIT_EDGES_IN_ORDER = 11;
    VISIT_VERTICES_IN_ORDER = 12;
    VISIT_SUBPATH = 13;
    NEGATE = 14;
    ZONE = 15;
  }
  
  optional Type type = 1;
  
  optional PBAvoidEdgeConstraint avoid_edge_constraint = 2;
  optional PBAvoidVertexConstraint avoid_vertex_constraint = 3;
  optional PBAndConstraint and_constraint = 4;
  optional PBOrConstraint or_constraint = 5;
  optional PBAvoidPathConstraint avoid_path_constraint = 6;
  optional PBDeprefEdgeConstraint depref_edge_constraint = 7;
  optional PBDeprefVertexConstraint depref_vertex_constraint = 8;
  optional PBVisitEdgeConstraint visit_edge_constraint = 9;
  optional PBVisitVertexConstraint visit_vertex_constraint = 10;
  optional PBVisitEdgesInOrderConstraint visit_edges_in_order_constraint = 11;
  optional PBVisitVerticesInOrderConstraint visit_vertices_in_order_constraint = 12;
  optional PBNegateConstraint negate_constraint = 14;
}

message PBNegateConstraint {
  optional PBConstraint constraint = 1;
}

message PBVisitVerticesInOrderConstraint {
  repeated string vertices = 1;
}

message PBVisitEdgesInOrderConstraint {
  repeated net.PBGraphLink edges = 1;
}

message PBVisitVertexConstraint {
  optional string id = 1;
}

message PBVisitEdgeConstraint {
  optional net.PBGraphLink edge = 1;
}

message PBDeprefVertexConstraint {
  optional string id = 1;
}

message PBDeprefEdgeConstraint {
  optional net.PBGraphLink edge = 1;
}

message PBAvoidPathConstraint {
  repeated net.PBGraphLink path = 1;
}

message PBAvoidEdgeConstraint {
  optional net.PBGraphLink edge = 1;
}

message PBAvoidVertexConstraint {
  optional string id = 1;
}

message PBAndConstraint {
  optional PBConstraint op_one = 1;
  optional PBConstraint op_two = 2;
}

message PBOrConstraint {
  optional PBConstraint op_one = 1;
  optional PBConstraint op_two = 2;
}

message PBDFSRequest {
  optional uint32 max_depth_hops = 1; 
  optional uint32 max_depth_metric = 2;
  optional uint32 max_duration_ms = 3;
  optional uint32 steps_to_check_for_stop = 4;
  optional string src = 5;

  // Whether or not to produce node-disjoint paths.
  // By default paths are edge-disjoint.
  optional bool node_disjoint = 6;
}

message SrcDstPairPathCacheStats {
  optional string src = 1;
  optional string dst = 2;

  // Total number of paths in the cache for this src/dst pair.
  optional uint32 num_paths = 3;

  // A 100 values representing the percentiles of the distribution of
  // path latencies in seconds.
  repeated double path_length_sec_distribution = 4;

  // A 100 values representing the percentiles of the distribution of
  // path lengths in hops.
  repeated uint32 path_length_hops_distribution = 5;
}

// Statistics about the path cache.
message PathCacheStats {
  // Total number of paths in the cache.
  optional uint64 num_paths = 1;

  // Estimate for the total number of bytes used by the cache.
  optional uint64 bytes_used = 2;

  // A 100 values representing the percentiles of the distribution of
  // path latencies in seconds. For all paths in the cache.
  repeated double path_length_sec_distribution = 3;

  // A 100 values representing the percentiles of the distribution of
  // path lengths in hops. For all paths in the cache.
  repeated uint32 path_length_hops_distribution = 4;
  
  // Statistics about the src/dst pairs.
  repeated SrcDstPairPathCacheStats src_dst_pair_stats = 5;
}